050-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,718 @@
+> In the single namespace case the cause is likely a mistake by a team, who need to know about the each of the conflicting resources to resolve the problem.
+> Support for multiple namesapces is motivated by a kind of multitenancy use case, where namespaces are assigned a disjoint subset of the topic space and management of those topics is delegated to different teams with their own namespace (e.g. using Kube RBAC). 
+> We can't assume that either team involved in the conflict know about the existence of the other.
+> Therefore a different `reason` seems appropriate.
+
+
+### Operational considerations
+
+#### Changing `metadata.name`
+
+This WILL be supported by:
+
+1. unmanaging the topic (`spec.managed=false`).
+2. deleting the `KafkaTopic`.
+3. recreating the `KafkaTopic` with a new name.
+
+#### Change the `metadata.namespace` 
+
+In the case of a UTO configured for multiple namesapces it will be possible to move the `KafkaTopic` for a topic in Kafka from one namespace to another by:
+
+  1. Unmanging the `KafkaTopic` in the old namespace (as described above)
+  2. Deleting the `KafkaTopic` from the old namespace,
+  3. Revising the policy (requiring a UTO restart)
+  4. Creating a new `KafkaTopic` in the new namespace.
+
+#### Changing `spec.topicName`
+
+This will NOT be supported.
+It will be detected by a mismatch between `spec.topicName` and `status.topicName`.
+Changes thus detected will result in an error condition.
+
+#### Decreasing `spec.partitions`
+
+This will NOT be supported, because it's not supported by Kafka.
+The user will either need to revert the `spec.partitions`, or recreate the topic in Kafka (which they can do via the `KafkaTopic` or not).
+
+#### Changing `spec.relicas`
+
+This will NOT be supported using as part of this proposal.
+
+#### Recreating topics in the Kafka cluster
+
+This WILL be supported.
+
+#### Deployment of application with `KafkaTopics`
+
+Consider an application deployed via a manifest containing a `Deployment` and some `KafkaTopics`.
+If the Kafka cluster is configured with `auto.create.topics.enable` there is a race condition between:
+
+1. the operator reacting to the new `KafkaTopics`.
+2. the application starting and triggering autocreation of the topics.
+
+**If the operator wins:** then auto-creation will not happen, and the user's intent is honoured.
+
+**If the application wins:** then the topics will be created using the default configuration for the cluster. The operator will later reconcile the `KafkaTopics` which will result in reconfiguration the default configuration for auto-created topics differs from the `KafkaTopic.spec`. That reconfiguration will either succeed, or fail (e.g. because a `KafkaTopic` has a lower `spec.partitions` than the auto-created topic).
+
+To avoid this it is **recommended** that the Topic Operator not be used with cluster where `auto.create.topics.enable` is `true`. 
+The operator will emit a `WARN` level log on startup if autocreation is enabled. 
```

Comment:
I've seen this happening with users more than one time and recovering from bad cluster defaults requires service disruption. In addition to the log on startup, can we also emit a Kube event with this warning?

Reply:
Do we know that users are more likely to notice a Kube `Event` than a warning in the logs?  

Emitting a Kube `Event` doesn't avoid any service disruption caused by changing the default cluster behaviour. It only makes the likely mistake more noticeable. And it's not _necessarily_ a mistake: It's only a problem if the user isn't creating their topics prior to application deployment. And even if it _is_ a mistake, it's not necessarily consequential (we expect the topic would get reconfigured almost immediately, so it's only a problem for certain changes, like RF change, or changing the #partitions). 

I'm not opposed to doing this, but I'm not completely convinced it would actually make a difference.

Reply:
> It's only a problem if the user isn't creating their topics prior to application deployment. 

Agree, but this is exactly the situation depicted in the gitops vs application race, which is fairly common nowadays. I think we could be fine with the log warning and a documentation paragraph/section describing potential issues with topic auto autocreation, maybe in relation to few common use cases.